File: backend/src/main/java/com/lingko/lingko/api/evaluation/dto/ScoreStatus.java
package com.lingko.lingko.api.evaluation.dto;

/**
 * 평가 응답에 실린 점수를 사용자에게 보여도 되는지 나타내는 전송 계약 상태값이다.
 *
 * 공급자 token이 기준 문장과 정렬되지 않으면 점수를 "0점"으로 내리는 대신 신뢰 불가로 표시해야
 * 하는데, nullable 숫자 하나로는 "0점"과 "모름"을 구분할 수 없어 상태값을 함께 둔다.
 * enum으로 정의해 서버 안에서는 오타와 누락된 분기를 컴파일 시점에 막고, JSON에는 Jackson 기본
 * 직렬화인 {@link Enum#name()} 그대로 나가 기존 wire format을 유지한다.
 */
public enum ScoreStatus {

    /** 점수 필드가 실제 측정값이며 사용자에게 노출해도 된다. */
    AVAILABLE,

    /** 점수를 신뢰할 수 없어 숫자를 노출하면 안 된다. 이때 점수 필드는 항상 null이다. */
    UNAVAILABLE;

    /**
     * 점수 유무만으로 상태를 정하는 저장·복원 경로에서 두 값의 대응을 한곳에 모은다.
     *
     * 호출부마다 삼항 연산자를 반복하면 null을 AVAILABLE로 잘못 매핑하는 실수가 생길 수 있어
     * 변환 규칙 자체를 이 타입이 소유한다.
     */
    public static ScoreStatus ofNullableScore(Integer score) {
        return score == null ? UNAVAILABLE : AVAILABLE;
    }

    public boolean isAvailable() {
        return this == AVAILABLE;
    }
}

File: app/lib/models/score_status.dart
// 파일 의도: 평가 점수를 사용자에게 노출해도 되는지를 나타내는 서버 상태값을 앱 타입으로 고정한다.
// 선택 이유: 문자열 비교가 화면마다 흩어지면 오타 하나가 조용히 "신뢰할 수 없는 점수 노출"로 이어진다.

/// 서버 `scoreStatus` 계약의 앱 표현이다.
///
/// 점수 필드가 null인 것만으로는 "0점"과 "측정하지 못함"을 구분할 수 없어 서버가 상태값을 함께 보낸다.
/// [unavailable]이면 점수는 항상 null이며, 화면은 숫자를 감추되 발음 가이드는 그대로 제공해야 한다.
enum ScoreStatus {
  available,
  unavailable;

  /// 서버 JSON 값을 앱 타입으로 변환하며, 아는 값이 아니면 [unavailable]로 닫는다.
  ///
  /// 앱은 서버와 동시에 배포되지 않으므로 구버전 앱이 새 상태값을 만나는 상황이 정상 경로다.
  /// 이때 모르는 값을 낙관적으로 해석하면 신뢰할 수 없는 점수를 사용자에게 보여주게 되므로,
  /// 판단이 서지 않는 모든 입력을 안전한 쪽으로 보내는 fail-closed 규칙을 택했다.
  static ScoreStatus fromWire(Object? raw) {
    return raw == 'AVAILABLE' ? ScoreStatus.available : ScoreStatus.unavailable;
  }

  /// 점수 유무만 아는 응답에서 상태를 유도한다. 과거 기록 API처럼 상태값이 없는 경우에 쓴다.
  static ScoreStatus ofNullableScore(Object? score) {
    return score == null ? ScoreStatus.unavailable : ScoreStatus.available;
  }

  bool get isAvailable => this == ScoreStatus.available;
}
